src/library/vm/vm_pos_info_provider.cpp

```diff
@@ -0,0 +1,61 @@
+/*
+Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+
+Author: Jared Roesch
+*/
+#include "library/vm/vm_pos_info.h"
+#include "kernel/pos_info_provider.h"
+#include "library/vm/vm_nat.h"
+#include "library/vm/vm_expr.h"
+#include "library/vm/vm_option.h"
+
+namespace lean {
+
+struct vm_pos_info_provider : public vm_external {
+    pos_info_provider const & m_val;
```

Comment:
This is terribly unsafe if this `vm_obj` outlives the parser, e.g. if its stored in an info_manager, or used asynchronously in a different task (all easily possible since ts_clone is implemented).

I think a much simpler and safer design would be to add a single new (impure) function `tactic.expr_pos : expr → tactic expr`, which just calls `get_pos_info_provider().get_pos_info_or_some(to_expr(...))`.  We can imagine that the pos_info_provider is passed via the tactic_state, and we don't need to pass it around.

Another alternative option is to store a parser_pos_info_provider here (maybe in a `std::shared_ptr<pos_info_provider const>` if you like an extra sprinkle of genericity).